Fix #370: Logger instances are lazily initialized

carterkozak opened the pull request:
Default java loggers are no longer initialized when custom
ExceptionHandler instances are provided. This allows logging
frameworks to delegate to disruptor without risking deadlocks.

Implementation notes:

This introduces a new public utility class `ExceptionHandlers` which exposes a singleton `ExceptionHandler<Object>` that can be lazily initialized. This allows the handler to be centrally controlled and shared between `BatchEventProcessor` and `ExceptionHandlerWrapper`.

The `ExceptionHandler` instances are stored in non-volatile fields, so I opted to null check and produce the default singleton instead of attempting to initialize the field to some default just prior to initialization. This way even unexpected concurrent initial accesses should not be able to produce different behavior if this patch is applied.

`BatchEventProcessor` is performance sensitive, so I opted to structure changes in such a way that method size will not grow, in case it's optimized by size for inlining. Exception handling generally occurs out of hot paths, so additional method call overhead should be negligible, and the calling method size should be slightly lower, avoiding the `exceptionHandler` object lookup bytecode. As an aside, the `BatchEventProcessor` hot loop may benefit from a single final field read outside of the loop to avoid unnecessary indirection because `-XX:+TrustFinalNonStaticFields` still isn't the default JVM behavior.

carterkozak commented:
This could be implemented with fewer lines of code, but the code would be less obvious and incur slightly higher startup cost (e.g. unnecessary object creation). I'd be happy to implement this in another way if you prefer. Thanks!

carterkozak commented:
This is the simpler alternative for comparison: https://github.com/carterkozak/disruptor/commit/e350a7fe309eaa149330d965eb86740b8e043eb0

Palmr commented:
Thanks for the PR @carterkozak all looks good.

It would be nice to have a test that fails if the loading of the logger in `FatalExceptionHandler` or elsewhere changed in future to be non-lazy.
But I'm not sure what such a test would look like and if it would be possible/sensible to do?

Did you have any thoughts on this?
I saw in the earlier issue there's a JMH benchmark which was used to try and catch a deadlock on the log4j side, I wondered if this could be of use? Perhaps should be a JCStress test?
Setup just in Disruptor might be too heavy though, as I assume it would need some kind of small test logger framework too?

carterkozak commented:
Thanks, @Palmr. I wanted to implement a simple unit test for this, but it's difficult to do in a reasonable way.

The JMH reproducer requires a logging framework which is both initialized via the system logger as well as an implementation which relies on disruptor.

The closest idea I had was to run a test with a custom delegating LogManager implementation (initialized on the first use) which tracks interactions, that way we can validate whether or not the system logger has been accessed. This approach relies on static state (loggers themselves tend to be static state) so the unit test would be required to run in its own fork, otherwise would potentially produce incorrect data. Also, junit 5 initializes a JUL logger in setup, which would invalidate the test. A unit test could reuse the current classpath and launch another JVM via a test `main` method, but this style test is fragile.
I can put some time into experimenting with such a test if you like, but it's likely to be difficult to maintain over time. What do you think?

Palmr commented:
I would advise doing this kind of test with JUnit or JMH involved for the reasons you've mentioned.

A JCStress test in Termination mode (e.g. https://github.com/openjdk/jcstress/blob/010fc224643490dd13519ede33eda9ff0c276fc3/jcstress-samples/src/main/java/org/openjdk/jcstress/samples/APISample_03_Termination.java) would potentially be a better test framework for writing this kind of test where we're looking for a fairly rare occurrence of a deadlock between threads.

Though as I mentioned before it may be a bit too heavyweight of a test to have in the Disruptor code, with the need for a custom delegating LogManager implementation and potential other setup required.

The patch seems fair, mostly unnoticeable in normal circumstances and in theory it should fix the problems initially mentioned in the issue.
A test would be nice to give me more tangible proof that it fixes the original issue, but I can run this patched disruptor through the original JMH-based reproducer to get my fix of seeing a test flip from red to green. (I take it you've already done this?)

I'm thinking I should merge this PR in its current state and leave any potential tests as a "nice to have" that could come in a later PR if time is found and results of experimentation with testing frameworks yields results.

carterkozak commented:
I've put together a working JCStress test here: #374
It's not all that bad, but it's a stretch to use JCStress in this case -- it's a convenient place to run validation on a forked JVM.

In the JCStress case I'm not attempting to reproduce the deadlock, but validate that the root cause of the deadlock (logger initialization when a Disruptor instance is created and started) no longer occurs. That PR fails on master, but passes when this commit is present.
Would you like me to add that commit to this PR?

Palmr commented:
> Would you like me to add that commit to this PR?

I think so.

I like the idea that there is _something_ that shows this commit does indeed fix the root cause of that deadlock.
If you add the test to this PR I'll merge it all, and can then close #374

carterkozak commented:
Sounds good, thanks! I've added that commit to this PR and added inline comments describing the `Unsafe` piece of the test which validates logging has not been initialized.

Palmr commented:
Seems the CI checks hadn't run until I allowed it to, not something I've seen GitHub do before.

But now it has run it seems you've fallen foul of recently-added checkstyle rules: missing a `@return` in your JavaDoc at `ExceptionHandlers.java:23`

Once that's added I can merge :)

carterkozak commented:
I see, after a rebase I was able to reproduce the checkstyle failure. Updated.

carterkozak commented:
Thanks for the review :-)
